File: PHP/345. Reverse Vowels of a String.php
<?php

/**
 * @noinspection PhpIllegalPsrClassPathInspection
 * @noinspection PhpMultipleClassDeclarationsInspection
 * @noinspection PhpUnused
 */
class Solution
{
    public function reverseVowels(string $s): string
    {
        $length = strlen($s);

        if ($length <= 1) {
            return $s;
        }

        $vowels = [
            'a' => true, 'e' => true, 'i' => true, 'o' => true, 'u' => true,
            'A' => true, 'E' => true, 'I' => true, 'O' => true, 'U' => true,
        ];

        $left = 0;
        $right = $length - 1;

        while ($left < $length) {
            if ($left >= $right) {
                break;
            }

            if (array_key_exists($s[$left], $vowels)) {
                while ($right > $left) {
                    if (array_key_exists($s[$right], $vowels)) {
                        [$s[$left], $s[$right]] = [$s[$right], $s[$left]];
                        $right--;
                        break;
                    }

                    $right--;
                }
            }

            $left++;
        }

        return $s;
    }
}

File: Python/345. Reverse Vowels of a String.py
class Solution:
    # noinspection PyMethodMayBeStatic
    def reverseVowels(self, s: str) -> str:
        length = len(s)

        if length <= 1:
            return s

        vowels = [
            'a', 'e', 'i', 'o', 'u',
        ]

        left = 0
        right = length - 1
        s = list(s)

        while left < length:
            if left >= right:
                break

            if s[left].lower() in vowels:
                while right > left:
                    if s[right].lower() in vowels:
                        s[left], s[right] = s[right], s[left]
                        right -= 1
                        break
                    right -= 1

            left += 1

        return ''.join(s)
